File: implementation/scipamato/scipamato-wicket/src/test/java/ch/difty/scipamato/web/AbstractPanelTest.java
package ch.difty.scipamato.web;

import static org.assertj.core.api.Assertions.*;

import org.apache.wicket.markup.html.form.TextField;
import org.apache.wicket.model.Model;
import org.junit.Test;

import ch.difty.scipamato.web.test.TestAbstractPanel;
import de.agilecoders.wicket.extensions.markup.html.bootstrap.form.checkboxx.CheckBoxX;

public class AbstractPanelTest extends WicketTest {

    @Test
    public void testViewMode_withOneArgConstructor() {
        assertViewMode(new AbstractPanel<TestRecord>("panel") {
            private static final long serialVersionUID = 1L;
        });
    }

    @Test
    public void testViewMode_withTwoArgConstructor() {
        assertViewMode(new AbstractPanel<TestRecord>("panel", Model.of(new TestRecord(1, "foo"))) {
            private static final long serialVersionUID = 1L;
        });
    }

    @Test
    public void testViewMode_withThreeArgConstructor() {
        assertViewMode(new AbstractPanel<TestRecord>("id", Model.of(new TestRecord(1, "foo")), Mode.VIEW) {
            private static final long serialVersionUID = 1L;
        });
    }

    private void assertViewMode(AbstractPanel<TestRecord> p) {
        assertThat(p.getMode()).isEqualTo(Mode.VIEW);
        assertThat(p.getSubmitLinkResourceLabel()).isEqualTo("button.disabled.label");
        assertThat(p.isViewMode()).isTrue();
        assertThat(p.isEditMode()).isFalse();
        assertThat(p.isSearchMode()).isFalse();
    }

    @Test
    public void testEditMode() {
        assertEditMode(new AbstractPanel<TestRecord>("panel", Model.of(new TestRecord(1, "foo")), Mode.EDIT) {
            private static final long serialVersionUID = 1L;
        });
    }

    private void assertEditMode(AbstractPanel<TestRecord> p) {
        assertThat(p.getMode()).isEqualTo(Mode.EDIT);
        assertThat(p.getSubmitLinkResourceLabel()).isEqualTo("button.save.label");
        assertThat(p.isViewMode()).isFalse();
        assertThat(p.isEditMode()).isTrue();
        assertThat(p.isSearchMode()).isFalse();
    }

    @Test
    public void testSearchMode() {
        assertSearchMode(new AbstractPanel<TestRecord>("panel", Model.of(new TestRecord(1, "foo")), Mode.SEARCH) {
            private static final long serialVersionUID = 1L;
        });
    }

    private void assertSearchMode(AbstractPanel<TestRecord> p) {
        assertThat(p.getMode()).isEqualTo(Mode.SEARCH);
        assertThat(p.getSubmitLinkResourceLabel()).isEqualTo("button.search.label");
        assertThat(p.isViewMode()).isFalse();
        assertThat(p.isEditMode()).isFalse();
        assertThat(p.isSearchMode()).isTrue();
    }

    @Test
    public void assertTestAbstractPanel_inViewMode() {
        getTester().startComponentInPage(new TestAbstractPanel("panel", Mode.VIEW));
        assertComponents();
    }

    private void assertComponents() {
        getTester().assertLabel("panel:fooLabel", "Foo");
        getTester().assertComponent("panel:foo", TextField.class);
        getTester().assertLabel("panel:barLabel", "Bar");
        getTester().assertComponent("panel:bar", TextField.class);
        getTester().assertLabel("panel:bazLabel", "Baz");
        getTester().assertComponent("panel:baz", CheckBoxX.class);
    }

}

File: common/common-wicket/src/test/kotlin/ch/difty/scipamato/common/web/AbstractPanelTest.kt
package ch.difty.scipamato.common.web

import de.agilecoders.wicket.extensions.markup.html.bootstrap.form.checkboxx.CheckBoxX
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeFalse
import org.amshove.kluent.shouldBeTrue
import org.apache.wicket.bean.validation.PropertyValidator
import org.apache.wicket.markup.html.form.FormComponent
import org.apache.wicket.markup.html.form.TextField
import org.apache.wicket.model.Model
import org.junit.jupiter.api.Test

internal class AbstractPanelTest : WicketBaseTest() {

    @Test
    fun testViewMode_withOneArgConstructor() {
        assertViewMode(object : AbstractPanel<TestRecord>("panel") {
            private val serialVersionUID: Long = 1L
        })
    }

    @Test
    fun testViewMode_withTwoArgConstructor() {
        assertViewMode(object : AbstractPanel<TestRecord>("panel", Model.of(TestRecord(1, "foo"))) {
            private val serialVersionUID: Long = 1L
        })
    }

    @Test
    fun testViewMode_withThreeArgConstructor() {
        assertViewMode(object : AbstractPanel<TestRecord>("id", Model.of(TestRecord(1, "foo")), Mode.VIEW) {
            private val serialVersionUID: Long = 1L
        })
    }

    private fun assertViewMode(p: AbstractPanel<TestRecord>) {
        p.mode shouldBeEqualTo Mode.VIEW
        p.submitLinkResourceLabel shouldBeEqualTo "button.disabled.label"
        p.isViewMode.shouldBeTrue()
        p.isEditMode.shouldBeFalse()
        p.isSearchMode.shouldBeFalse()
    }

    @Test
    fun testEditMode() {
        assertEditMode(object : AbstractPanel<TestRecord>("panel", Model.of(TestRecord(1, "foo")), Mode.EDIT) {
            private val serialVersionUID: Long = 1L
        })
    }

    private fun assertEditMode(p: AbstractPanel<TestRecord>) {
        p.mode shouldBeEqualTo Mode.EDIT
        p.submitLinkResourceLabel shouldBeEqualTo "button.save.label"
        p.isViewMode.shouldBeFalse()
        p.isEditMode.shouldBeTrue()
        p.isSearchMode.shouldBeFalse()
    }

    @Test
    fun testSearchMode() {
        assertSearchMode(object : AbstractPanel<TestRecord>("panel", Model.of(TestRecord(1, "foo")), Mode.SEARCH) {
            private val serialVersionUID: Long = 1L
        })
    }

    private fun assertSearchMode(p: AbstractPanel<TestRecord>) {
        p.mode shouldBeEqualTo Mode.SEARCH
        p.submitLinkResourceLabel shouldBeEqualTo "button.search.label"
        p.isViewMode.shouldBeFalse()
        p.isEditMode.shouldBeFalse()
        p.isSearchMode.shouldBeTrue()
    }

    @Test
    fun assertTestAbstractPanel_inViewMode() {
        tester.startComponentInPage(TestAbstractPanel("panel", Mode.VIEW))
        assertComponents()
    }

    private fun assertComponents() {
        tester.assertLabel("panel:fooLabel", "Foo")
        tester.assertComponent("panel:foo", TextField::class.java)
        tester.assertLabel("panel:barLabel", "Bar")
        tester.assertComponent("panel:bar", TextField::class.java)
        tester.assertLabel("panel:bazLabel", "Baz")
        tester.assertComponent("panel:baz", CheckBoxX::class.java)
    }

    @Test
    fun queuingFieldAndLabel_withPropertyValidatorInEditMode_addsLatterToComponent() {
        val fc = mockk<FormComponent<String>>(relaxed = true) {
            every { id } returns "fcId"
        }
        val pv = mockk<PropertyValidator<FormComponent<String>>>()

        val p = TestAbstractPanel("panel", Mode.EDIT)
        p.queueFieldAndLabel(fc, pv)

        verify(exactly = 3) { fc.id }
        verify { fc.label = any() }
        verify { fc.setOutputMarkupId(true) }
        verify { fc.add(pv) }
        confirmVerified(fc, pv)
    }

    @Test
    fun queuingFieldAndLabel_withPropertyValidatorInViewMode_addsNothingToComponent() {
        val fc = mockk<FormComponent<String>>(relaxed = true) {
            every { id } returns "fcId"
        }
        val pv = mockk<PropertyValidator<FormComponent<String>>>()

        val p = TestAbstractPanel("panel", Mode.VIEW)
        p.queueFieldAndLabel(fc, pv)

        verify(exactly = 3) { fc.id }
        verify { fc.label = any() }
        verify { fc.setOutputMarkupId(true) }
        verify(exactly = 0) { fc.add(pv) }
        confirmVerified(fc, pv)
    }

    @Test
    fun queuingFieldAndLabel_withNullPropertyValidator_addsNothingToComponent() {
        val fc = mockk<FormComponent<String>>(relaxed = true) {
            every { id } returns "fcId"
        }

        val p = TestAbstractPanel("panel", Mode.EDIT)
        p.queueFieldAndLabel(fc, null)

        verify(exactly = 3) { fc.id }
        verify { fc.setOutputMarkupId(true) }
        verify { fc.label = any() }
        confirmVerified(fc)
    }
}
